ServerCore/SyncController.cs

```diff
@@ -0,0 +1,565 @@
+        /// </summary>
+        public class AnnotationRequest
+        {
+            public readonly int Key;
+            public readonly string Contents;
+
+            public AnnotationRequest(int key, string contents)
+            {
+                Key = key;
+                Contents = contents;
+            }
+        }
+
+        /// <summary>
+        ///   The DecodedSyncRequest class represents a request to perform a sync, decoded from the format posted
+        ///   by Javascript.
+        /// </summary>
+        public class DecodedSyncRequest
+        {
+            /// <summary>
+            ///   The list of puzzled IDs to check whether they're solved yet.  Could be null.
+            /// </summary>
+
+            public readonly List<int> QueryPuzzleIds;
+
+            /// <summary>
+            ///   If MinSolveCount isn't null, the requester wants any pieces the user has earned
+            ///   corresponding to a solve count greater than or equal to MinSolveCount.  For
+            ///   instance, if this is the first time the requester is asking for information, it'll
+            ///   set this to 0; if this is a subsequent time, the requester will set it to
+            ///   MaxSolveCount+1, where MaxSolveCount was the maximum solve count they were told
+            ///   they received last time.  If this is an "annotation upload only" sync, then
+            ///   MinSolveCount will be null, meaning "don't bother computing what pieces I've
+            ///   earned, because I don't want any right now".
+            /// </summary>
+            public readonly int? MinSolveCount;
+
+            /// <summary>
+            ///   If AnnotationRequests isn't null, the requester wants to upload all the
+            ///   annotations described by these requests.
+            /// </summary>
+            public readonly List<AnnotationRequest> AnnotationRequests;
+
+            /// <summary>
+            ///   If LastSyncTime isn't null, the requester wants all the annotations that the team
+            ///   has created for this puzzle since LastSyncTime.  If LastSyncTime is null, the
+            ///   requester wants all annotations that the team has created for this puzzle, no
+            ///   matter when they were created.
+            /// </summary>
+            public readonly DateTime? LastSyncTime;
+
+            /// <summary>
+            ///   If an error was encountered while processing the raw parameters passed to the sync
+            ///   controller, this will be a non-null string that can be returned to the user to
+            ///   give a meaningful error message.  It's intended that if the request comes from a
+            ///   page we created, no error will ever be raised.
+            /// </summary>
+            public readonly string DecodingError;
+
+            public DecodedSyncRequest(List<int> query_puzzle_ids, int? min_solve_count, string annotations, string last_sync_time,
```

Comment:
Why is this doing custom deserialization rather than just using Newtonsoft Json to deserialize the whole class?

Reply:
HTTP allows you to POST a set of variables, some of which are optional, so I'm taking advantage of that.  Essentially, my API has a set of inputs, and I use a separate POST variable for each of them.

Reply:
Also, I don't want to worry about different browser implementations of Javascript encoding DateTimes differently, so I do all the serialization/deserialization of DateTimes on the server.